Library/Formula/mlton.rb

```diff
@@ -0,0 +1,131 @@
+
+# MLton is a self-hosting compiler for Standard ML.
+# In order to build from source, this formula uses the corresponding
+# upstream binary release to bootstrap.
+
+class Mlton < Formula
+  desc "Whole-program, optimizing compiler for Standard ML"
+  homepage "http://mlton.org"
+  url "https://downloads.sourceforge.net/project/mlton/mlton/20130715/mlton-20130715.src.tgz"
+  version "20130715"
+  sha256 "215857ad11d44f8d94c27f75e74017aa44b2c9703304bcec9e38c20433143d6c"
+
+  depends_on "gmp"
+
+  # The corresponding upstream binary release used to bootstrap.
+  resource "bootstrap" do
+    url "https://downloads.sourceforge.net/project/mlton/mlton/20130715/mlton-20130715-2.amd64-darwin.gmp-static.tgz"
+    sha256 "16a6d4e300f45f4af094692cf8033390e4634fa4c072caf6e9c288234100ad22"
+  end
+
+  # The upstream source release hard codes common locations for GMP;
+  # on darwin, these include '/opt/local' (MacPorts) and '/sw' (Fink).
+  # Replace these with placeholders for the Homebrew GMP location.
+  patch :DATA
+
+  def install
+    # Finish configuring source release with Homebrew GMP location.
+    inreplace "bin/mlton-script", "gmpHomebrewCCOpts=''", "gmpHomebrewCCOpts='-I#{Formula["gmp"].opt_prefix}/include'"
+    inreplace "bin/mlton-script", "gmpHomebrewLinkOpts=''", "gmpHomebrewLinkOpts='-L#{Formula["gmp"].opt_prefix}/lib'"
+    inreplace "runtime/Makefile", "GMP_HOMEBREW_XCFLAGS :=", "GMP_HOMEBREW_XCFLAGS := -I#{Formula["gmp"].opt_prefix}/include"
+
+    # Import the corresponding upstream binary release to 'bootstrap'.
+    bootstrap = buildpath/"bootstrap"
+    bootstrap_bin = bootstrap/"usr/local/bin"
+    bootstrap_lib = bootstrap/"usr/local/lib"
+    mkdir bootstrap
+    resource("bootstrap").stage(bootstrap/"usr")
+    # The binary release assumes that the MLton support files are in
+    # '/usr/local/lib/mlton' and that GMP is in '/usr/local',
+    # '/opt/local' (MacPorts), or '/sw' (Fink).
+    # Would like to patch/inreplace the
+    # 'bootstrap/usr/local/bin/mlton' script as is done above for the
+    # source release, but the Homebrew patch API does not support
+    # patching external resources.  Since the binary and source
+    # releases correspond, the 'bootstrap/usr/local/bin/mlton' script
+    # is identical to the (original, un-patched/un-inreplaced)
+    # 'bin/mlton-script', modulo a 'lib=...' that determines the
+    # location of the MLton support files.  Therefore, use the
+    # (patched/inreplaced) 'bin/mlton-script' for the
+    # 'bootstrap/usr/local/bin/mlton' script, updating 'lib=...' to
+    # the location of the bootstrap MLton support files.
+    cp "bin/mlton-script", bootstrap_bin/"mlton"
+    inreplace bootstrap_bin/"mlton", /^lib=.*/, "lib='#{bootstrap_lib}/mlton'"
+    chmod "a+x", bootstrap_bin/"mlton"
+    ENV.prepend_path "PATH", bootstrap_bin
+
+    ENV.deparallelize
```

Comment:
Regarding the `inreplace` changes: I'm not sure what would be submitted upstream.  This is patching the upstream binary release, which assumes that it will be unpacked in `/` (yielding, `/usr/local/lib/mlton` for the support files).  Again, the [install instructions](http://mlton.org/Installation) explicitly mention editing the `lib` variable in the `mlton` script if the binary package is installed somewhere else.

Regarding the deparallelize: I can certainly open an upstream issue.  However, a previous attempt at making the MLton build parallel safe determined that the patch would be quite invasive; see [this message](https://sourceforge.net/p/mlton/mailman/message/31031131/).